### PR TITLE
BAS Platform Extension

### PR DESCRIPTION
**Pull request by perspectivus1**
Added a bas-platform extension that exposes an API and interfaces:
    1. Classes for creating actions
    1. Method for executing actions
    1. Method for for obtaining vscode extension APIs

In the future, we'll add more functionality. For example, a message queue for contributor-to-manager communications.

The guided-development extension has an `extensionDepedencies` to the bas-platform extension. It uses it for performing actions.

Each contributing extension also has an `extensionDepedencies` to the bas-platform extension. These extensions use it for creating actions and for obtaining the guided-development extension's API. An example of proper use can be found in the `simple-contrib` sample extension.

The `sample-action-client` extension demonstrates the use of the bas-platform API to perform **scheduled actions**.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -38,10 +38,13 @@ jobs: # a collection of steps
       - store_artifacts: # special step to save test results as as artifact
           path: frontend/coverage
           destination: frontend-coverage
+      - run:
+          name: Change Permissions
+          command: sudo chown -R $(whoami) /usr/local
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply .circleci/config.yml:43]: For `npm link`. We might want to keep performing `npm link` for CI, even once the dependent packages are available on `npm`.

File: .circleci/config.yml [hunk 2]
```diff
@@ -38,10 +38,13 @@ jobs: # a collection of steps
...
       - run:
           name: build backend
           working_directory: backend
-          command: npm run backend
+          command: npm run backend-dev
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply .circleci/config.yml:47]: So we can install, build and link to the dependent `bas-platform` package

File: backend/src/extension.ts [hunk 3]
```diff
@@ -27,9 +28,12 @@ export function activate(context: vscode.ExtensionContext) {
 	registerAndSubscribeCommand("guidedDevelopment.toggleOutput", guidedDevelopmentPanel.toggleOutput.bind(guidedDevelopmentPanel));
 	registerWebviewPanelSerializer(guidedDevelopmentPanel);
 
-	const vscodeEvents = new VSCodeEvents();
-	api.setSetData(vscodeEvents, vscodeEvents.setData);
-	return api.default;
+	const vscodeEvents = VSCodeEvents.getInstance();
+	const basAPI: typeof bas = await vscode.extensions.getExtension("SAPOSS.bas-platform")?.exports;
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply backend/src/extension.ts:32]: Please remind me what the limitation is in Theia regarding case of extension IDs. Would the line above not work in Theia?

File: bas-platform/src/api.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+                    }
+                }, 500);
+            }
+        });
+    
+        return promise;    
+    },
+
+    actions: {
+        performAction,
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply bas-platform/src/api.ts:28]: The `bas.actions` object exposes all actions APIs, including classes and methods.

Perhaps all classes should be on the root of the `bas` API hierarchy, similar to the way it's done in `vscode`... @idoprz, what do you think?

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply .circleci/config.yml:47]: but this is only for development. We don't want/need it in CI.

File: .circleci/config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply .circleci/config.yml:43]: If he dependent packages are available on npm then I don't want to perform any 'npm link'

File: backend/src/extension.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/src/extension.ts:32]: Would not work in Theia.

File: README.md [hunk 5]
```diff
@@ -30,6 +30,26 @@ To test run the framework you only need to build and install the backend package
...
     ```
+
+#### Temporarily
+... until `bas-platform` and `guided-development` packages are published to npm:
+
+In folder `bas-platform`, perform `npm link`.
+
+In folder `backend`, perform `npm link`.
+
+For each of the following folders, perform `npm link @sap-devx/bas-platform` and `npm link @sap-devx/guided-development`:
```
Comment (tomer-epstein) [inline README.md:41]: `npm link bas-platform` and `npm link guided-development`

File: README.md [hunk 6]
```diff
@@ -30,6 +30,26 @@ To test run the framework you only need to build and install the backend package
...
+* `contrib-cake`
+* `contrib-oven`
+* `vscode-contrib1`
+* `vscode-contrib2`
+* `vscode-contrib3`
+* `simple-contrib`
+
+For each of the following folders, perform `npm link @sap-devx/bas-platform`:
```
Comment (tomer-epstein) [inline README.md:49]: `npm link bas-platform`

File: .circleci/config.yml [hunk 7]
```diff
@@ -38,13 +38,10 @@ jobs: # a collection of steps
...
           path: frontend/coverage
           destination: frontend-coverage
-      - run:
-          name: Change Permissions
-          command: sudo chown -R $(whoami) /usr/local
       - run:
           name: build backend
           working_directory: backend
-          command: npm run backend-dev
+          command: npm run backend
```
Comment (tomer-epstein) [inline .circleci/config.yml]: 👍

File: backend/package.json [hunk 8]
```diff
@@ -112,11 +110,13 @@
 		"watch": "tsc -watch -p ./",
 		"package": "npm run webpack && vsce package .",
 		"test": "nyc mocha -p tsconfig.json --opts ./mocha.opts",
+		"test2": "nyc mocha -p tsconfig.json --opts ./mocha.opts",
```
Comment (tomer-epstein) [inline backend/package.json:113]: ??

File: backend/tsconfig.json [hunk 9]
```diff
@@ -23,10 +23,10 @@
 		"target": "es6",
 		"traceResolution": false,
 		"declaration": true,
-		"alwaysStrict": true
+		"alwaysStrict": true,
```
Comment (tomer-epstein) [inline backend/tsconfig.json:26]: "alwaysStrict": true

File: backend/tsconfig.json [hunk 10]
```diff
@@ -23,10 +23,10 @@
...
 	},
 	"include": [
 		"./src/**/*",
 		"./tests/**/*"
-	]
+	],
```
Comment (tomer-epstein) [inline backend/tsconfig.json:31]: ]

File: guided-development-types/package.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    "type": "git",
+    "url": "https://github.com/SAP/guided-development.git",
+    "directory": "guided-development-types"
+  },
+  "main": "./out/src/types/GuidedDev",
+  "types": "./out/src/types/GuidedDev",
+  "scripts": {
+    "prep": "cd ../backend && npm i && npm run compile",
+    "copy": "ncp ../backend/out out",
+    "zprepublishOnly": "npm run prep && npm run copy"
```
Comment (tomer-epstein) [inline guided-development-types/package.json:18]: why z ?

**Review by tomer-epstein — CHANGES_REQUESTED** [review]
Also we need to update the README. not to perform npm link.

File: README.md (hunk 5 above)
Comment (tomer-epstein) [inline README.md:41]: remove, pls.

File: README.md (hunk 6 above)
Comment (tomer-epstein) [inline README.md:49]: remove, pls.

File: contrib-oven/src/types/GuidedDev.ts [hunk 12]
```diff
@@ -26,46 +26,6 @@ export interface IItem {
     labels: {[key:string]:string}[];
 }
 
-export interface IAction {
-    name: string;
-    title?: string;
-}
```
Comment (tomer-epstein) [inline contrib-oven/src/types/GuidedDev.ts:32]: file should be deleted.

Comment (tomer-epstein) [inline contrib-oven/src/types/GuidedDev.ts:32]: file should be deleted.

**Review by tomer-epstein — CHANGES_REQUESTED** [review]

**Review by tomer-epstein — APPROVED** [review]

**Pull request merged**